pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+    def f() -> (async (int, str) -> bool): pass
+
+    def f() -> async (int) -> async (str) -> bool: pass
+    def f() -> (async (int) -> (async (str) -> bool)): pass
+
+
+Trailing Commas
+~~~~~~~~~~~~~~~
+
+- Following the precedent of function signatures, putting a comma in
+  an empty arguments list is illegal, ``(,) -> bool`` is a syntax
+  error.
+- Again following precedent, trailing commas are otherwise always
+  permitted::
+
+
+    ((int,) -> bool == (int) -> bool
+    ((int, **P,) -> bool == (int, **P) -> bool
+    ((...,) -> bool) == ((...) -> bool)
+
+Allowing trailing commas also gives autoformatters more flexibility
+when splitting callable types across lines, which is always legal
+following standard python whitespace rules.
+
+
+Disallowing ``...`` as an Argument Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under normal circumstances, any valid expression is permitted where we
+want a type annotation and ``...`` is a valid expression. This is
+never semantically valid and all type checkers would reject it, but
+the grammar would allow it if we did not explicitly prevent this.
+
+We decided that there were compelling reasons to prevent it: - The
+semantics of ``(...) -> bool`` are different from ``(T) -> bool`` for
+any valid type T: ``(...)`` is a special form indicating
+``AnyArguments`` whereas ``T`` is a type parameter in the arguments
+list.  - ``...`` is used as a placeholder default value to indicate an
+optional argument in stubs and Callback Protocols. Allowing it in the
+position of a type could easily lead to confusion and possibly bugs
+due to typos.
+
+Since ``...`` is meaningless as a type and there are usability
+concerns, our grammar rules it out and the following is a syntax
+error::
+
+    (int, ...) -> bool
+
+Incompatibility with other possible uses of ``*`` and ``**``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any
+future proposal using a bare ``**<some_type>`` to type
+``kwargs``. This seems acceptable because:
+
+- If we ever do want such a syntax, it would be clearer to require an
+  argument name anyway. This would also make the type look more
+  similar to a function signature. In other words, if we ever support
+  typing ``kwargs`` in callable types, we would prefer ``(int,
+  **kwargs: str)`` rather than ``(int, **str)``.
+- PEP 646 unpacking syntax would rule out using ``*<some_type>`` for
+  ``args``. The ``kwargs`` case is similar enough that this rules out
+  a bare ``**<some_type>`` anyway.
+
+
+
+Compatibility with Arrow-Based Lambda Syntax
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To the best of our knowledge there is no active discussion of
+arrow-style lambda syntax that we are aware of, but it is nonetheless
+worth considering what possibilities would be ruled out by adopting
+this proposal.
+
+It would be incompatible with this proposal to adopt the same a
```

Comment:
```suggestion
It would be incompatible with this proposal to adopt a
```

Reply:
Can you explain why? (Because `(int) -> str` is a valid expression.)

-------

That reminds me: we need to have a subsection for aliases (`Foo = (int) -> str`, generic aliases) and how a callable type behaves at runtime when used as an expression. 

@gvanrossum  might have thoughts on things we should cover (especially from python-dev's point of view).